### DISCUSSION TITLE
Force unicode support in dialect to avoid `test <XYZ>`

### DISCUSSION THREAD

**Discussion by nicolaerosia** [sqlalchemy/sqlalchemy#5893] [Usage Questions] [answered]
Everytime I connect to a Vertica database the driver issues the following queries:
```
SELECT CAST('test plain returns' AS VARCHAR(60)) as anon_1
SELECT CAST('test unicode returns' AS VARCHAR(60)) as anon_1
```

Is there a way to disable these?
I'm using sqlalchemy 1.3.21 (conda) and [sqlalchemy-vertica-python](https://github.com/bluelabsio/sqlalchemy-vertica-python)

grep'ed the source code and I've found some reference about this in `changelog/migration_13.rst`, `convert_unicode parameters deprecated`
The sqlalchemy-vertica-python dialect is based on `PGDialect` from `sqlalchemy.dialects.postgresql.base`
It seems to set something related to `unicode` here, https://github.com/bluelabsio/sqlalchemy-vertica-python/blob/master/sqla_vertica_python/vertica_python.py#L27
```
    supports_unicode_statements = True
    supports_unicode_binds = True
    supports_native_decimal = True
```
but maybe it's not what is required?

**Comment by zzzeek** [comment, marked as answer]
Here's what I can give you:

1. the queries above are not on every connection, they are only on a per-engine basis.   Assuming you are using create_engine() properly, which is that the engine is created [just once, at the module level](https://docs.sqlalchemy.org/en/13/core/connections.html#basic-usage), you would only see these queries emitted once per application run, not once per connection.


2. The above mentioned queries are [removed in SQLAlchemy 1.4 when Python 3 is in use](https://docs.sqlalchemy.org/en/14/changelog/changelog_14.html#change-d3650b27f4c852ac687ea55326051ed0) which was accomplished by #5315 .

In SQLAlchemy 1.3, the only way to prevent this query on first connect is to override the _check_unicode_returns() method, but this method continues to be required for DBAPIs that are running under Python 2, as Python 2 DBAPIs are all over the map as far as what they return, and the technique of "run some data through" to check was actually a huge leap forward in stability, compared to the previous approach of hardcoding and assuming.

**Reply by nicolaerosia** [reply]
Thank you for the reply, I'm using 32 processes that connect to the server, each doing create_engine because I think I'm not supposed to connect only in the master process that starts them, right?

**Reply by zzzeek** [reply]
use a single engine and call dispose() at the start of a process.  See option 2 at https://docs.sqlalchemy.org/en/13/core/pooling.html#pooling-multiprocessing .   but also, these queries are of negligible overhead in any case compared to that of setting up the new connection overall.

**Reply by nicolaerosia** [reply]
@zzzeek Thank you, I'm now using SQLAlchemy 1.4.0b2, with `future=True` to be ready for 2.0 :)
I have a different issue though, maybe you can give me a hint. I find that INSERTs are slow, and after profiling, I noticed that I get back the primary keys in `inserted_primary_key_rows` but I don't need them. I'm using the core API, similar to this:
```
sql_connection.execute(
    sqlalchemy.insert(
        MyTable,
        returning=None,
        return_defaults=False,
    ),
    records,
)
```

The `records` is a list of dictionaries which contain all values, hence I don't need the inserted primary keys.

where `MyTable` is:
```
class MyTable(Base):
    key1 = sqlalchemy.Column(
        sqlalchemy.Integer, nullable=False, primary_key=True
    )
    key2 = sqlalchemy.Column(sqlalchemy.DateTime, nullable=False, primary_key=True)

    data1 = sqlalchemy.Column(sqlalchemy.Float, nullable=False)
```

Is there a way to disable the returns?

In the tutorial I see that it's enabled by default, https://docs.sqlalchemy.org/en/14/tutorial/data.html#insert-returning
```
[...]
The RETURNING clause for supported backends is used automatically in order to retrieve the last inserted primary key value as well as the values for server defaults.
[...]
```

**Reply by nicolaerosia** [reply]
@zzzeek sorry for the noise, I think I found what I need, `inline()` in ` sqlalchemy.sql.expression.Insert`

**Reply by nicolaerosia** [reply]
@zzzeek, actually I see that it doesn't work, here's what I tried:
1) add `"implicit_returning": False,` to `__table_args__` in `MyTable(Base)`
2) add `implicit_returning=False` to `sqlalchemy.create_engine(...)`
3) add `inline=True` to `sqlalchemy.insert` / `sqlalchemy.insert.inline()` tried both, no luck

The values are still populated in `sql_res.inserted_primary_key_rows` and `sql_res.rowcount`
```
sql_connection.engine.dialect.implicit_returning
False
sql_connection.engine.dialect.full_returning
True
```

Any hints are appreciated :)

**Reply by zzzeek** [reply]
insert.inline() will prevent an INSERT statement from doing anything special to get the primary key, here's an [MCVE](http://stackoverflow.com/help/mcve):
```python

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import String
from sqlalchemy import Table


m = MetaData()
t = Table('t', m, Column('id', Integer, primary_key=True), Column('data', String))

e = create_engine("postgresql://scott:tiger@localhost/test", echo='debug')


with e.begin() as conn:
    m.create_all(conn)

    result = conn.execute(
        t.insert(),
        {"data": "some data where we get the PK"}
    )
    print(result.inserted_primary_key)

    result = conn.execute(
        t.insert().inline(),
        {"data": "some data where we dont get the primary key"}
    )
    print(result.inserted_primary_key)

```

output with the relevant SQL:

```sql
INSERT INTO t (data) VALUES (%(data)s) RETURNING t.id
{'data': 'some data where we get the PK'}
Col ('id',)
Row (4,)
(4,)
INSERT INTO t (data) VALUES (%(data)s)
{'data': 'some data where we dont get the primary key'}
(None,)

```

additionally, INSERTs are only "efficient" if you run them as "executemany" style, that is, pass many rows at once.  RETURNING will not be used for this form either:

```python
with e.begin() as conn:
    m.create_all(conn)

    result = conn.execute(
        t.insert(),
        [
            {"data": "some data %s" % i}
            for i in range(100)
        ]
    )

```


SQL:

```sql
INSERT INTO t (data) VALUES (%(data)s)
2021-02-10 09:45:55,005 INFO sqlalchemy.engine.Engine [generated in 0.00019s] ({'data': 'some data 0'}, {'data': 'some data 1'}, {'data': 'some data 2'}, {'data': 'some data 3'}, {'data': 'some data 4'}, {'data': 'some data 5'}, {'data': 'some data 6'}, {'data': 'some data 7'}  ... displaying 10 of 100 total bound parameter sets ...  {'data': 'some data 98'}, {'data': 'some data 99'})

```

in the latter example, SQLAlchemy 1.4 will also make automatic use of [psycopg2 execute_values()](https://www.psycopg.org/docs/extras.html#psycopg2.extras.execute_values) which will rewrite the INSERT as a single statement with many VALUES elements, but it still will not use RETURNING unless return_defaults() was requested.

the "inserted_primary_key_rows" collection is "populated", but with a None-value:

```
[(None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), (None,), ...

```

the generation of this useless list is a little bit non-performant, can you clarify the specific source of the "slowness"?  do you have a cprofile to share?

**Reply by zzzeek** [reply]
.rowcount is also not relevant for an INSERT, that is just accessing cursor.rowcount which is always there and is per spec supposed to be the value "-1", but it's whatever the DBAPI puts there.

**Reply by zzzeek** [reply]
I've made it so that "empty pk" thing does not run unless you access that member:  89dc4562adb38367ee5fabbcc04ee44968af4906

**Reply by nicolaerosia** [reply]
thanks a lot! After profiling it seems that IDs are actually got from `_setup_ins_pk_from_empty` which just gets them from compiled parameters, so I assume there is no transfer from the server.

seems like time is wasted in `vertica_python/vertica/connection.py:624(read_message)`, not in the actual transfer to server.
The weird thing is that the server reports a runtime of ~15miliseconds (Vertica server)
https://github.com/vertica/vertica-python/blob/master/vertica_python/vertica/connection.py#L624
https://github.com/vertica/vertica-python/blob/master/vertica_python/vertica/cursor.py#L273
https://github.com/vertica/vertica-python/blob/master/vertica_python/vertica/cursor.py#L431

```
select start_timestamp, request_duration_ms, request
FROM query_requests
WHERE
start_timestamp > sysdate - 1/24
AND
request_type = 'LOAD'
ORDER BY start_timestamp DESC;
```

```
         37780 function calls in 2.385 seconds

   Ordered by: cumulative time
   List reduced from 145 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    2.385    2.385 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/sqlalchemy/future/engine.py:254(execute)
        1    0.000    0.000    2.385    2.385 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1482(_execute_20)
        1    0.000    0.000    2.385    2.385 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/sqlalchemy/sql/elements.py:309(_execute_on_connection)
        1    0.000    0.000    2.385    2.385 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1317(_execute_clauseelement)
        1    0.000    0.000    2.385    2.385 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1559(_execute_context)
        1    0.000    0.000    2.378    2.378 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/sqlalchemy/engine/default.py:677(do_executemany)
        1    0.000    0.000    2.378    2.378 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/vertica_python/vertica/cursor.py:226(executemany)
        1    0.000    0.000    2.353    2.353 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/vertica_python/vertica/cursor.py:397(copy)
        4    0.000    0.000    2.352    0.588 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/vertica_python/vertica/connection.py:624(read_message)
       12    0.000    0.000    2.352    0.196 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/vertica_python/vertica/connection.py:686(read_bytes)
        6    2.352    0.392    2.352    0.392 {method 'recv' of '_socket.socket' objects}
        1    0.000    0.000    0.025    0.025 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/vertica_python/vertica/cursor.py:264(<listcomp>)
      360    0.006    0.000    0.024    0.000 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/vertica_python/vertica/cursor.py:554(format_operation_with_parameters)
     1800    0.002    0.000    0.009    0.000 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/re.py:203(sub)
     1800    0.005    0.000    0.007    0.000 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/vertica_python/vertica/cursor.py:518(object_to_string)
     1800    0.002    0.000    0.005    0.000 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/re.py:289(_compile)
        1    0.000    0.000    0.004    0.004 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/sqlalchemy/engine/default.py:1338(_setup_result_proxy)
        1    0.000    0.000    0.004    0.004 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/sqlalchemy/engine/default.py:1402(_setup_dml_or_text_result)
        1    0.000    0.000    0.004    0.004 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/sqlalchemy/engine/default.py:1476(_setup_ins_pk_from_empty)
        1    0.001    0.001    0.004    0.004 /home/nicolaerosia/miniconda3/envs/devel_py38/lib/python3.8/site-packages/sqlalchemy/engine/default.py:1480(<listcomp>)
```

**Reply by gordthompson** [reply]
Hi @nicolaerosia - If you need more help with https://github.com/vertica/vertica-python please ask them.
